### ISSUE TITLE
`IntegrityError` thrown, blaming `collPrepUniqueId` when adding a duplicate barcode

### ISSUE THREAD

**Issue by grantfitzsimmons** [specify/specify7#2886]
> As mentioned in the call we have been finding that Specify will often crash when trying to add additional preparations to an existing collection object.
> 
> The report 2023-01-17 happened when a digitiser was trying to add a preparation to a record they had databased through the minimal data entry script. They then opened up the newly created collection object to add a preparation for the second sheet.
> 
> The report 2023-01-26 was triggered by me trying to add an additional preparation to a collection object that was migrated into Specify. 

It looks like the underlying issue is related to the fact that a preparation has a unique identifier that is already being used:

```
  "responseText": "{\"exception\": \"IntegrityError\", \"message\": 1062, \"data\": \"Duplicate entry '4-E01004897' for key 'collPrepUniqueId'\"
```

​Specify should catch this before throwing an ugly error that loses the entered data.

There already is a barcode with the value of `E01004897` in the database.

[Specify 7 Crash Report - 2023-01-26T16 24 31.648Z.txt](https://github.com/specify/specify7/files/10511414/Specify.7.Crash.Report.-.2023-01-26T16.24.31.648Z.txt)

**Reported By:** Robyn at RBGE

**Comment by melton-jason** [comment]
This uniqueness rule does not exist on either the front end or the backend, which would explain the Integrity Error

**Comment by melton-jason** [comment]
Interestingly, I noticed some discrepancies between specify databases. It appears the `collPrepUniqueId` key is a fragment of some previous schema change before 6.8.01 (unless there is some setting which alters the tables). 

Here are the keys (indexes) for a local database created in 6.8.02 (The keys are identical for a database created in 6.8.01). I have emphazised the the related keys to this issue. 

![specify_preparation_keys](https://user-images.githubusercontent.com/64045831/215563997-0974a401-bca5-4d40-a529-a485812b8ffa.png)

We see that the BarCode column is in two keys, CollectionMemberID and PrepBarCodeIDX. 

Things become a little more clear when we view the statement which created these keys:

![specify_preparation_table](https://user-images.githubusercontent.com/64045831/215564442-9f9aaf95-849e-4b02-933b-6ace4c0fa9c2.png)

The only UNIQUE key that BarCode is associated with is `CollectionMemberID`. There is no reference whatsoever to a 'collPrepUniqueId' key. 

On the contrary, here is a list of keys for the preparation table of a local copy of aafc-dao (local copies of vimsfish and umich also shared this behavior):

![aafc_preparation_keys](https://user-images.githubusercontent.com/64045831/215565790-d6f78af8-534c-4526-97f0-8c45158a6920.png)

If we likewise view the statement which created the table:

![aafc_preparation_table](https://user-images.githubusercontent.com/64045831/215565929-05884a1b-74bf-4443-909e-d3ab16ff28af.png)

(Note that these database also have `KEY 'PrepBarCodeIdx' ('BarCode')`, which was exlcuded from the screenshot to preserve screen space)

Why is there a discrepancy between databases?

**Comment by maxxxxxdlp** [comment]
Could have been a manual modification or a failed schema migration. For confidence, could you create a new database using SpWizard and check the database schema?

**Comment by melton-jason** [comment]
> Could have been a manual modification or a failed schema migration. For confidence, could you create a new database using SpWizard and check the database schema?

The created database shows the same schema as my testing local instance. That is, there is no `collPrepUniqueId` and the BarCode field is included in a `CollectionMemberID` unique key.

**Comment by maxxxxxdlp** [comment]
From @grantfitzsimmons:

> <img width="1423" alt="image" src="https://user-images.githubusercontent.com/37256050/206347457-5beaddc0-36f6-40ae-aee9-4083fee387f9.png">
>
> [Specify 7 Crash Report - 2022-12-08T03_16_55.589Z.txt](https://github.com/specify/specify7/files/10181769/Specify.7.Crash.Report.-.2022-12-08T03_16_55.589Z.txt)
>
> This issue occurred when adding a new preparation to the record with a barcode that is already in use. You can recreate by creating a new prep with the barcode ``123`.
> 
> https://herbrbge6802-edge.test.specifysystems.org/specify/view/collectionobject/151675/

**Comment by grantfitzsimmons** [comment]
This is still really annoying **and** destructive.

`(1062, "Duplicate entry '4-E01199145' for key 'collPrepUniqueId'")`

**Comment by grantfitzsimmons** [comment]
Still happening on `v7.9-dev`

**Comment by grantfitzsimmons** [comment]
We probably should have fixed this in https://github.com/specify/specify7/pull/3996 rather than in https://github.com/specify/specify7/pull/3790. What do you think @melton-jason?

**Comment by melton-jason** [comment]
Fixed as of #4215
